ls1intum/Artemis
# `Discussion`: Add search and filtering to exercise and lecture discussion tabs

### Checklist
#### General


#### Client

### Motivation and Context


When users are using the discussion section of exercises or lectures, in particular, long lists of posts make it hard to reach information. Furthermore while creating a new Post in such a context, users have to scroll to the bottom of the long post list, since the "New Post" button is located at the very bottom of the list.

Solves #5050.

### Description

To let the users reach information in a more convenient way, we enable searching and filtering of posts while on the page of a specific exercise or lecture. Furthermore, a second "New Post" is added before the beginning of the list, if there are more than 2 Posts present.

### Steps for Testing

Prerequisites:
- 1 Student
- 1 course with postings enabled

1. Log in to Artemis
2. Navigate to Course discussion page
3. Make sure that all functionality is still present as before. 
4. Go to the page of a specific exercise or lecture.
5. Make sure that searching and filtering work as expected.
6. Make sure that a single "Create Post" button is present if there are less than 3 posts.
7. Make sure that two "Create Post" buttons are present if there are more than 2 posts. 

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
course-discussion.component.ts  | 80.24% | 95.5% |
discussion-section.component.ts | 86% |  90.78% |
course-discussion.directive.ts | 91.66% | 100% |
metis.util.ts | 100% | 100% |
posting-thread.component.ts  | 100% | 100% |


### Screenshots


![image](https://user-images.githubusercontent.com/67344158/170739861-5e136a68-b16d-4d8d-b9d1-a67570b3e88a.png)
![image](https://user-images.githubusercontent.com/67344158/170739898-233ba9b6-a36f-4d41-9d3a-a42e67c9e061.png)
![image](https://user-images.githubusercontent.com/67344158/170739952-5d89df3a-3ffc-4dc7-9144-5f702ba43d81.png)
